From 9a490c1b34bf4e42582ee2f6b180c83ab26bab21 Mon Sep 17 00:00:00 2001
From: "Gu, Wangyi" <wangyi.gu@intel.com>
Date: Tue, 21 Aug 2012 09:37:56 +0800
Subject: [PATCH] [PORT FROM MAIN]Update the rotation information before
 playing new clips

The rotation degree information must be updated when
playing several video clips one by one in video editor.
The clips use the same ANativewindow to render video frames.
If the first rendered clip have 90, 180 or 270 rotation
degrees, the following clips will also use the this
information to rotate. This is not right. Rotation degree
must be updated before rendering frames of a new video clips.
Original BZ: 33132

DO NOT UPSTREAM!

Change-Id: Ib00e6e0777bae489237384e9ef933c0310bbb800
Category: aosp improvement
Domain: Video.Media-video editor
Origin: 33132
Signed-off-by: Gu, Wangyi <wangyi.gu@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>
Signed-off-by: chuzhoux <chux.zhou@intel.com>

--- a/libvideoeditor/lvpp/NativeWindowRenderer.cpp
+++ b/libvideoeditor/lvpp/NativeWindowRenderer.cpp
@@ -317,6 +317,11 @@ NativeWindowRenderer::~NativeWindowRenderer() {
 void NativeWindowRenderer::render(RenderInput* input) {
     sp<GLConsumer> ST = input->mST;
     sp<Surface> STC = input->mSTC;
+    int err;
+    err = native_window_set_buffers_transform(STC.get(), input->mTransform);
+    if (err != 0) {
+        ALOGE("native_window_set_buffers_transform failed!");
+    }
 
     if (input->mIsExternalBuffer) {
         queueExternalBuffer(STC.get(), input->mBuffer,
@@ -603,6 +616,14 @@ void RenderInput::updateVideoSize(sp<MetaData> meta) {
         mWidth = mHeight;
         mHeight = tmp;
     }
+
+    switch (rotationDegrees) {
+        case 0: mTransform= 0; break;
+        case 90: mTransform = HAL_TRANSFORM_ROT_90; break;
+        case 180: mTransform = HAL_TRANSFORM_ROT_180; break;
+        case 270: mTransform = HAL_TRANSFORM_ROT_270; break;
+        default: mTransform = 0; break;
+    }
 }
 
 void RenderInput::render(MediaBuffer* buffer, uint32_t videoEffect,
diff --git a/libvideoeditor/lvpp/NativeWindowRenderer.h b/libvideoeditor/lvpp/NativeWindowRenderer.h
index 26b4cba..79b096e 100755
--- a/libvideoeditor/lvpp/NativeWindowRenderer.h
+++ b/libvideoeditor/lvpp/NativeWindowRenderer.h
@@ -167,6 +167,7 @@ private:
     sp<GLConsumer> mST;
     sp<Surface> mSTC;
     int mWidth, mHeight;
+    uint32_t mTransform;
 
     // These are only valid during render() calls
     uint32_t mVideoEffect;
